src/main/java/ch/njol/skript/expressions/ExprXYZComponent.java

```diff
@@ -0,0 +1,198 @@
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  Skript is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Skript.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright Peter Güttinger, SkriptLang team and contributors
+ */
+package ch.njol.skript.expressions;
+
+import org.bukkit.event.Event;
+import org.bukkit.util.Vector;
+import org.eclipse.jdt.annotation.Nullable;
+import org.joml.Quaternionf;
+
+import ch.njol.skript.Skript;
+import ch.njol.skript.classes.Changer;
+import ch.njol.skript.classes.Changer.ChangeMode;
+import ch.njol.skript.classes.Changer.ChangerUtils;
+import ch.njol.skript.doc.Description;
+import ch.njol.skript.doc.Examples;
+import ch.njol.skript.doc.Name;
+import ch.njol.skript.doc.Since;
+import ch.njol.skript.expressions.base.SimplePropertyExpression;
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.util.Kleenean;
+import ch.njol.util.coll.CollectionUtils;
+
+/**
+ * Ported by Sashie from skript-vectors with bi0qaw's permission.
+ * @author bi0qaw
+ */
+@Name("Vector/Quaternion - XYZ Component")
+@Description({
+	"Gets or changes the x, y or z component of <a href='classes.html#vector'>vectors</a>/<a href='classes.html#quaternion'>quaternions</a>.",
+	"cannot use w of vector. w is for quaternion only."
+})
+@Examples({
+	"set {_v} to vector 1, 2, 3",
+	"send \"%x of {_v}%, %y of {_v}%, %z of {_v}%\"",
+	"add 1 to x of {_v}",
+	"add 2 to y of {_v}",
+	"add 3 to z of {_v}",
+	"send \"%x of {_v}%, %y of {_v}%, %z of {_v}%\"",
+	"set x component of {_v} to 1",
+	"set y component of {_v} to 2",
+	"set z component of {_v} to 3",
+	"send \"%x component of {_v}%, %y component of {_v}%, %z component of {_v}%\""
+})
+@Since("2.2-dev28, INSERT VERSION (Quaternions)")
+public class ExprXYZComponent extends SimplePropertyExpression<Object, Number> {
+
+	static {
+		String types = "vectors";
+		if (Skript.isRunningMinecraft(1, 19, 4))
+			types += "/quaternions";
+		register(ExprXYZComponent.class, Number.class, "[vector|quaternion] (0¦w|1¦x|2¦y|3¦z) [component[s]]", types);
+	}
+
+	private final static Character[] axes = new Character[] {'w', 'x', 'y', 'z'};
+
+	private int axis;
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		axis = parseResult.mark;
+		return super.init(exprs, matchedPattern, isDelayed, parseResult);
+	}
+
+	@Override
+	public Number convert(Object object) {
+		if (object instanceof Vector) {
+			Vector vector = (Vector) object;
+			return axis == 1 ? vector.getX() : (axis == 2 ? vector.getY() : vector.getZ());
```

Comment:
This allows `w` to return the z component of a vector.

Reply:
yes. w could return null then?

Reply:
Yeah that'd be what I would expect to happen.